### PR TITLE
[Merged by Bors] - feat: Popoviciu's inequality

### PR DESCRIPTION
**Pull request by tukamilano**
Formalize [Popoviciu's inequality on variances](https://en.wikipedia.org/wiki/Popoviciu%27s_inequality_on_variances) through the [Bhatia-Davis inequality](https://en.wikipedia.org/wiki/Bhatia%E2%80%93Davis_inequality).

Co-authored-by: Yuma Mizuno mizuno.y.aj@gmail.com

### CODE REVIEW

File: Mathlib/Probability/Variance.lean
Comment (RemyDegenne) [inline Mathlib/Probability/Variance.lean:8]: That new import is unused. Please remove it.
The build failed in CI. If you click on that, you will get to a page that tells you why (this unused import is the reason).

Comment (tukamilano) [reply Mathlib/Probability/Variance.lean]: Thank you for your detailed instruction. I removed the new import.

File: Mathlib/Probability/Variance.lean
Comment (tukamilano) [reply Mathlib/Probability/Variance.lean]: Thank you!

File: Mathlib/Probability/Variance.lean
Comment (YaelDillies) [inline Mathlib/Probability/Variance.lean:349]: Do we really want to turn every deterministic statement ever into a statement about ae?

File: Mathlib/Probability/Variance.lean
Comment (YaelDillies) [inline Mathlib/Probability/Variance.lean:360]:
```suggestion

/-- ** Popoviciu's inequality on variance**

The variance of a random variable `X` satisfying `a ≤ X ≤ b`  almost everywhere is at most
`((b - a) / 2) ^ 2`. -/
```

File: Mathlib/Probability/Variance.lean
Comment (yuma-mizuno) [reply Mathlib/Probability/Variance.lean:349]: > Do we really want to turn every deterministic statement ever into a statement about ae?

Would it be nice to have a tactic that generates this?

Comment (YaelDillies) [reply Mathlib/Probability/Variance.lean:349]: It's called `filter_upwards`, as witnessed by the code

Comment (yuma-mizuno) [reply Mathlib/Probability/Variance.lean:349]: Yes indeed, but I meant something like `attribute to_ae` that generates `abs_le_max_abs_abs_ae` from `abs_le_max_abs_abs`.

File: Mathlib/Probability/Variance.lean
Comment (RemyDegenne) [inline Mathlib/Probability/Variance.lean:378]: here is a second example of parentheses and types that can be removed. Please make similar changes elsewhere.
```suggestion
      _ = ∫ ω, - X ω ^ 2 + (a + b) * X ω ∂μ - ∫ _, a * b ∂μ :=
```

File: Mathlib/Probability/Variance.lean
Comment (RemyDegenne) [inline Mathlib/Probability/Variance.lean:400]: Why do you introduce a `have : ∀ x` to prove something for all `x`, then use it for one particular value? You could simply prove it for that value. and gain 3 lines.

Or you could extract the `have` as an independent lemma.

File: Mathlib/Probability/Variance.lean
Comment (RemyDegenne) [inline Mathlib/Probability/Variance.lean:353]: Having those two counditions for ae bounded seems a bit verbose. What about this:
```suggestion
theorem memℒp_of_mem_Icc [IsFiniteMeasure μ]
    {a b : ℝ} {X : Ω → ℝ} (h : ∀ᵐ ω ∂μ, X ω ∈ Icc a b)
    (hX : AEMeasurable X μ) (p : ENNReal) : Memℒp X p μ :=
```
Also that lemma should be moved to another file. `#find_home!` is useful to find a good file.

**Review by RemyDegenne — COMMENTED** [review]
Thanks for the PR!

File: Mathlib/Probability/Variance.lean
Comment (YaelDillies) [reply Mathlib/Probability/Variance.lean:349]: Sorry, I was unclear. I mean that we do *not* want to turn every deterministic statement into an ae statement. Instead, we should use `filter_upwards` whenever needed in later proofs. Here this means *not* proving `abs_le_max_ae` but instead calling `filter_upwards` wherever you used `abs_le_max_ae`.

File: Mathlib/Probability/Variance.lean
Comment (tukamilano) [reply Mathlib/Probability/Variance.lean:379]: Thank you for your suggestion about parentheses. I'll revise it.

File: Mathlib/Probability/Variance.lean
Comment (tukamilano) [reply Mathlib/Probability/Variance.lean:401]: Thank you for pointing that out.
I wasn't able to clean up the code completely.

File: Mathlib/Probability/Variance.lean
Comment (tukamilano) [reply Mathlib/Probability/Variance.lean]: > Having those two counditions for ae bounded seems a bit verbose. What about this:

I realized two bounded conditions are verbose, but I was not sure which way is the most clear, so thank you for telling me the answer.

> Also that lemma should be moved to another file. `#find_home!` is useful to find a good file.

I moved memℒp_of_mem_Icc to Mathlib.MeasureTheory.Function.LpSeminorm.Basic file and added the import.

File: Mathlib/Probability/Variance.lean
Comment (YaelDillies) [inline Mathlib/Probability/Variance.lean:353]: Can you deduce this from the [Bhatia-Davis inequality](https://en.wikipedia.org/wiki/Bhatia%E2%80%93Davis_inequality)? It can be stated as
```
/-- **The Bhatia-Davis inequality on variance**

The variance of a random variable `X` satisfying `a ≤ X ≤ b`  almost everywhere is at most
`(b - 𝔼 X) * (𝔼 X - a)`. -/
lemma variance_le_sub_mul_sub [IsProbabilityMeasure μ] {a b : ℝ} {X : Ω → ℝ}
    (h : ∀ᵐ ω ∂μ, X ω ∈ Set.Icc a b) (hX : AEMeasurable X μ) :
    variance X μ ≤ (b - μ[X]) * (μ[X] - a) :=
```

File: Mathlib/Probability/Variance.lean
Comment (YaelDillies) [inline Mathlib/Probability/Variance.lean:347]:
```suggestion
/-- **Popoviciu's inequality on variance**
```

File: Mathlib/Probability/Variance.lean
Comment (tukamilano) [reply Mathlib/Probability/Variance.lean]: I divided the variance_square_bounded lemma into two lemmas in order for the Bhatia-Davis inequality to be available to apply.

File: Mathlib/MeasureTheory/Function/LpSeminorm/Basic.lean
Comment (YaelDillies) [inline Mathlib/MeasureTheory/Function/LpSeminorm/Basic.lean:732]:
```suggestion
    (by filter_upwards [ha, hb] with x using abs_le_max_abs_abs : ∀ᵐ x ∂μ, |f x| ≤ max |a| |b|)
```

File: Mathlib/Probability/Variance.lean
Comment (YaelDillies) [inline Mathlib/Probability/Variance.lean:401]: This nested calc is weird. Why not
```suggestion
  calc
    _ ≤ (b - μ[X]) * (μ[X] - a) := variance_le_sub_mul_sub h hX
    _ = ((b - a) / 2) ^ 2 - (μ[X] - (b + a) / 2) ^ 2 := by ring
    _ ≤ ((b - a) / 2) ^ 2 := sub_le_self _ (sq_nonneg _)
```

File: Mathlib/Probability/Variance.lean
Comment (yuma-mizuno) [inline Mathlib/Probability/Variance.lean:389]:
```suggestion
The variance of a random variable `X` satisfying `a ≤ X ≤ b` almost everywhere is at most
```

**Review by YaelDillies — APPROVED** [review]
I've shaved off 9 lines here and there. I think this is ready to go. Thanks again! and please keep on requesting my reviews 😁

maintainer merge

File: Mathlib/Probability/Variance.lean
Comment (RemyDegenne) [inline Mathlib/Probability/Variance.lean:384]: Perhaps the following name would fit better with our naming conventions:
```suggestion
lemma variance_le_sq_of_bounded [IsProbabilityMeasure μ] {a b : ℝ} {X : Ω → ℝ}
```
Or `variance_le_square_of_bounded` if you prefer.
Please change it also in the file docstring.

**Review by RemyDegenne — APPROVED** [review]
Thanks again!
Is the new import actually necessary? The bot says that there are no added imports with this PR, which makes me think that the file was perhaps already transitively imported.
bors d+
(please sort out those last comments, wait for CI to be green, then merge yourself with bors r+)

**Comment by RemyDegenne** [comment]
bors r-
Please update the file docstring with the new theorem name

**Comment by RemyDegenne** [comment]
Thanks!
bors r+

**Pull request closed without merging**
